cdc/entry/mounter.go

```diff
@@ -399,84 +408,108 @@ func (m *mounterImpl) mountRowKVEntry(tableInfo *model.TableInfo, row *rowKVEntr
 			TableID:     row.PhysicalTableID,
 			IsPartition: tableInfo.GetPartitionInfo() != nil,
 		},
-		Columns:         cols,
-		PreColumns:      preCols,
-		IndexColumns:    tableInfo.IndexColumnsOffset,
-		ApproximateSize: dataSize,
+		Columns:             cols,
+		PreColumns:          preCols,
+		IndexColumns:        tableInfo.IndexColumnsOffset,
+		ApproximateDataSize: dataSize,
 	}, nil
 }
 
 var emptyBytes = make([]byte, 0)
 
-func formatColVal(datum types.Datum, tp byte) (value interface{}, warn string, err error) {
+const (
+	sizeOfEmptyBytes  = int(unsafe.Sizeof(emptyBytes))
+	sizeOfEmptyString = int(unsafe.Sizeof(""))
+	sizeOfEmptyDatum  = int(unsafe.Sizeof(types.Datum{}))
+)
+
+func sizeOfDatumValue(d types.Datum) int {
+	array := [...]types.Datum{d}
+	return int(types.EstimatedMemUsage(array[:], 1)) - sizeOfEmptyDatum
+}
+
+func formatColVal(datum types.Datum, tp byte) (
+	value interface{}, size int, warn string, err error,
+) {
 	if datum.IsNull() {
-		return nil, "", nil
+		return nil, 0, "", nil
 	}
 	switch tp {
 	case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeNewDate, mysql.TypeTimestamp:
-		return datum.GetMysqlTime().String(), "", nil
+		v := datum.GetMysqlTime().String()
+		return v, len(v) + sizeOfEmptyString, "", nil
 	case mysql.TypeDuration:
-		return datum.GetMysqlDuration().String(), "", nil
+		v := datum.GetMysqlDuration().String()
+		return v, len(v) + sizeOfEmptyString, "", nil
 	case mysql.TypeJSON:
-		return datum.GetMysqlJSON().String(), "", nil
+		v := datum.GetMysqlJSON().String()
+		return v, len(v) + sizeOfEmptyString, "", nil
 	case mysql.TypeNewDecimal:
-		v := datum.GetMysqlDecimal()
-		if v == nil {
-			return nil, "", nil
+		d := datum.GetMysqlDecimal()
+		if d == nil {
+			return nil, 0, "", nil
 		}
```

Comment:
if `d` is a typed nil, the memory size of it is not zero?

Reply:
Yes, nil takes 0 byte.